### PR TITLE
[RFC]component: move pcm params from component to buffer struct

### PR DESCRIPTION
**Pull request by bkokoszx**
Moving pcm params from component to buffer is the first
step to improve parameter propagation within pipelines (#1280).
This commit moves several pcm paremeters (framet_fmt, buffer_fmt,
rate, channels, chmap, except direction) to comp_buffer struct.
Some of specific parameters are moved to private host and
kpb data.

Signed-off-by: Bartosz Kokoszko <bartoszx.kokoszko@linux.intel.com>

### CODE REVIEW

**Comment by lgirdwood** [comment]
@bkokoszx good work. Needs a little more update for CI, but I guess this is still Work in Progress.

**Comment by bkokoszx** [comment]
@lgirdwood 
Ok, CI has been updated. I would be grateful for review.

File: src/drivers/intel/cavs/ssp.c [hunk 1]
```diff
@@ -967,13 +967,13 @@ const struct dai_driver ssp_driver = {
...
-		.probe			= ssp_probe,
-		.remove			= ssp_remove,
+		.trigger = ssp_trigger,
+		.set_config = ssp_set_config,
+		.pm_context_store = ssp_context_store,
+		.pm_context_restore = ssp_context_restore,
+		.get_handshake = ssp_get_handshake,
+		.get_fifo = ssp_get_fifo,
+		.probe = ssp_probe,
+		.remove = ssp_remove,
```
Comment (paulstelian97) [inline src/drivers/intel/cavs/ssp.c]: Pass-by coding style change? Should be separate commit, although I'm not sure this particular change is even needed.

File: src/audio/pipeline.c [hunk 2]
```diff
@@ -255,13 +258,58 @@ static int pipeline_comp_period_frames(struct comp_dev *current)
...
+}
+
+/* save params changes made by component */
+static void pipeline_update_buffer_pcm_params(struct comp_buffer *buffer,
+					      void *data)
+{
+	struct sof_ipc_stream_params *params;
+	int i;
+
+	params = (struct sof_ipc_stream_params *)data;
```
Comment (lyakh) [inline src/audio/pipeline.c]: superfluous type-cast

File: src/include/sof/audio/buffer.h [hunk 3]
```diff
@@ -65,6 +67,13 @@ struct comp_buffer {
 	void (*cb)(void *data, uint32_t bytes);
 	void *cb_data;
 	int cb_type;
+
+	/* runtime stream params */
+	uint32_t frame_fmt;	/**< enum sof_ipc_frame */
+	uint32_t buffer_fmt;	/**< enum sof_ipc_buffer_format */
+	uint32_t rate;
+	uint16_t channels;
+	uint16_t chmap[SOF_IPC_MAX_CHANNELS];	/**< channel map - SOF_CHMAP_ */
```
Comment (lyakh) [inline src/include/sof/audio/buffer.h]: Do *all* of these really have to be fixed-width fields? E.g. rate? Wouldn't (unsigned) int do for it? Or channels? If formats are really those enums, why not also declare them as such?

File: src/include/sof/audio/buffer.h [hunk 4]
```diff
@@ -215,6 +225,77 @@ static inline void *buffer_get_frag(struct comp_buffer *buffer, void *ptr,
...
+	switch (buf->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		return 2 * buf->channels;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+	case SOF_IPC_FRAME_FLOAT:
+		return 4 * buf->channels;
+	default:
+		return 0;
+	}
```
Comment (lyakh) [inline src/include/sof/audio/buffer.h]: Couldn't this function just do `return buffer_sample_bytes(buf) * buf->channels;` ?

File: src/include/sof/audio/component.h [hunk 5]
```diff
@@ -246,7 +248,7 @@ struct comp_dev {
 	uint16_t state;		   /**< COMP_STATE_ */
 	uint64_t position;	   /**< component rendering position */
 	uint32_t frames;	   /**< number of frames we copy to sink */
-	uint32_t output_rate;      /**< 0 means all output rates are fine */
+	uint32_t output_rate;	   /**< 0 means all output rates are fine */
```
Comment (lyakh) [inline src/include/sof/audio/component.h]: more unrelated drive-by changes above

File: src/drivers/intel/cavs/ssp.c (hunk 1 above)
Comment (bkokoszx) [reply src/drivers/intel/cavs/ssp.c]: I've made it accidentally. It is corrected now.

File: src/include/sof/audio/buffer.h (hunk 4 above)
Comment (bkokoszx) [reply src/include/sof/audio/buffer.h]: I've just rebase PR with master. It's ```return frame_bytes(buf->frame_fmt, buf->channels);``` now.

File: src/include/sof/audio/component.h (hunk 5 above)
Comment (bkokoszx) [reply src/include/sof/audio/component.h]: You mean that all those cosding style changes should be in separate commit?

File: src/audio/pipeline.c (hunk 2 above)
Comment (bkokoszx) [reply src/audio/pipeline.c]: @lyakh 
Updated

**Comment by lgirdwood** [comment]
@mengdonglin @fkaczmar this is one PR that we will need to further validate after CI passes.

File: src/include/sof/audio/buffer.h (hunk 3 above)
Comment (bkokoszx) [reply src/include/sof/audio/buffer.h]: @lyakh 
I've just moved it to other struct. I do not mind changing these types to enum. 
@lgirdwood 
Is there a specific reason why we used fixed-width fields instead of e.g. enums?

**Comment by bkokoszx** [comment]
@lgirdwood 
PR has been updated. CI passes.

**Comment by zrombel** [comment]
Due to PR#2029 merged to master na CI bit depth conversion test change this PR must be rebased.

File: src/include/sof/audio/component.h (hunk 5 above)
Comment (paulstelian97) [reply src/include/sof/audio/component.h]: @bkokoszx Yup, they shouldn't be mixed with actual functional changes.

**Comment by kfialkox** [comment]
We've validate this PR with slim driver and python tests. There is no regression.

File: src/include/sof/audio/component.h [hunk 6]
```diff
@@ -189,11 +190,12 @@ struct comp_ops {
...
 
 	/** set component audio stream parameters */
-	int (*params)(struct comp_dev *dev);
+	int (*params)(struct comp_dev *dev,
+		      struct sof_ipc_stream_params *params);
 
 	/** set component audio stream parameters */
 	int (*dai_config)(struct comp_dev *dev,
-		struct sof_ipc_dai_config *dai_config);
+			  struct sof_ipc_dai_config *dai_config);
```
Comment (paulstelian97) [inline src/include/sof/audio/component.h]: Pass-by coding style change/fix, didn't check if it's a separate commit. Reminding myself to check this once I finish looking at everything else.

Comment (paulstelian97) [reply src/include/sof/audio/component.h]: Ah, it is indeed separate. We're good then.

**Comment by wenqingfu** [comment]
For some reason, there was no on-device testing performed on this PR. Seems it's causing massive regression on APL/BYT platforms with "aplay: set_params:1363: Unable to install hw params:" error.

**Comment by xiulipan** [comment]
@wenqingfu The test has **RFC** tag, and CI will skip the on-device test if we have RFC. Maybe we should not skip the on-device test even if the PR is RFC?

**Comment by plbossart** [comment]
We should do a revert on this PR, there's no reason why it should prevent all other PRs from being tested.

**Comment by bkokoszx** [comment]
@xiulipan @plbossart 
Sorry, I didn't know that PR's with RFC flag do not trigger device tests. I have prepared fix for #2154 - PR  #2156 - I think we do not have to revert PR #2130.

**Comment by wenqingfu** [comment]
> @wenqingfu The test has **RFC** tag, and CI will skip the on-device test if we have RFC. Maybe we should not skip the on-device test even if the PR is RFC?

The correct way should be developer to remove the RFC or WIP tag after they are happy with the PR. The policy does demonstrated its value in other cases, but it's new. So, let's do some propaganda work to let people realize this, then it'll be fine.
